File: exercism/python/space-age/space_age.py

class SpaceAge(object):

    ORBITAL_PERIODS = [(k, v * 31557600) for k, v in (
        ('earth', 1.0),
        ('mercury', 0.2408467),
        ('venus', 0.61519726),
        ('mars', 1.8808158),
        ('jupiter', 11.862615),
        ('saturn', 29.447498),
        ('uranus', 84.016846),
        ('neptune', 164.79132)
    )]

    def __init__(self, seconds):
        self.seconds = seconds
        for planet, p in self.ORBITAL_PERIODS:
            calc = lambda p=p: round(seconds / p, 2)
            setattr(self, 'on_' + planet, calc)

File: exercism/javascript/space-age/space-age.js


const planetYears = [
  ['Earth', 1],
  ['Mercury', 0.2408467],
  ['Venus', 0.61519726],
  ['Mars', 1.8808158],
  ['Jupiter', 11.862615],
  ['Saturn', 29.447498],
  ['Uranus', 84.016846],
  ['Neptune', 164.79132]
]

const secondsPerYear = 31557600

const toTwo = (n) => Math.round(n * 1e2) / 1e2

class SpaceAge {
  constructor (seconds) {
    this.seconds = seconds
    planetYears.forEach(([planet, earthYears]) => {
      this[`on${planet}`] = () =>
        toTwo(seconds / (secondsPerYear * earthYears))
    })
  }
}

module.exports = SpaceAge
